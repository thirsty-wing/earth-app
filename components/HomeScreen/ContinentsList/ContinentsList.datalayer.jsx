import useFetchEarth from "../../../utils/hooks/useFetchEarth";
import ContinentsList from "./ContinentsList";


/**
 * List of continents as pressables with data
 *
 * @returns {JSX.Element}
 */
function ContinentsListDatalayer() {

  const [
    data,
    responseStatus,
  ] = useFetchEarth(
    `query{
      continents {
        name
        code
      }
    }`
  );

  return(
    <ContinentsList
      data={data}
      responseStatus={responseStatus}
    />
  );
}

export default ContinentsListDatalayer;
