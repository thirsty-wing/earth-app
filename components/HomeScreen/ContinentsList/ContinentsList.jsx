import { Text, View, ActivityIndicator, Pressable } from "react-native";
import { theme } from "../../../utils/theme";
import { useRouter } from "expo-router";


/**
 * List of continents as pressables
 *
 * @param {object} props
 * @param {object} [props.data]
 * @param {"LOADING"|"ERROR"|"OK"} [props.responseStatus = "OK"]
 */
function ContinentsList(props) {

  const {
    data,
    responseStatus = "OK",
  } = props;

  const router = useRouter();

  /**
   * creates a function to be called in the continent Pressable
   *
   * @param {string} code
   */
  function makeDoNavigate(code) {
    return () => router.push(`continent/${code}`);
  }

  return(
    <View
      style={{
        flex: 1,
        gap: theme.spacing(),
      }}
    >

      { responseStatus === "LOADING" &&
        <ActivityIndicator size="large"/>
      }


      { responseStatus === "ERROR" &&
        <Text>Something went wrong</Text>
      }


      { responseStatus === "OK" &&
        data?.continents?.map(continent => (
          <Pressable
            key={`continent-${continent.code}`}
            style={{
              backgroundColor: theme.palette.background.offColor,
              borderRadius: 16,
              padding: 16
            }}
            onPress={makeDoNavigate(continent.code)}
          >
            <Text style={{ color: theme.palette.text.primary }}>
              {continent.name}
            </Text>
            <Text style={{ color: theme.palette.text.primary }}>
              {continent.code}
            </Text>
          </Pressable>
        ))
      }

    </View>
  );
}

export default ContinentsList;
