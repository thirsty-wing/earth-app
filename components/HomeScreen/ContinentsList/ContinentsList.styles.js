import { StyleSheet } from "react-native";

import { theme } from "../../../utils/theme";


const styles = StyleSheet.create({
  rootView: {
    flex: 1,
    gap: theme.spacing(),
  },
  continentPressable: {
    backgroundColor: theme.palette.action.primary,
    borderRadius: 16,
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    padding: 16,
  },
  continentName: {
    color: theme.palette.text.primary,
    fontSize: theme.sizes.text.subsection,
    flexWrap: "wrap",
  },
  continentCode: {
    color: theme.palette.text.primary,
    fontSize: theme.sizes.text.plain,
  },
  svgView: {
    backgroundColor: theme.palette.text.primary,
    borderRadius: 16,
    height: 125,
    width: 125,
    padding: theme.spacing(),
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
});

export default styles;
