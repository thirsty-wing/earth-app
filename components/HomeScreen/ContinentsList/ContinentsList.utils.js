/**
 * creates a function to be called in the continent Pressable
 *
 * @param {object} args
 * @param {object} args.navigation
 * @param {function} args.navigation.navigate
 * @param {string} args.code
 */
export function yieldDoNavigate(args) {

  const {
    navigation,
    code,
  } = args;

  return () => navigation.navigate("Continent", { code });
}
