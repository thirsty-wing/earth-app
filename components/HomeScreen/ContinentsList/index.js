import ContinentsList from "./ContinentsList.datalayer";

export default ContinentsList;
