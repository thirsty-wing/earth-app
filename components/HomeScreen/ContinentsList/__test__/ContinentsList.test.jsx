import ContinentsList from "../ContinentsList";
import { shallow } from "../../../../utils/test/shallow";


it("should render with continents", () => {
  const content = shallow(<ContinentsList
    data={{
      continents: [
        {
          name: "Pangea",
          code: "PG",
        },
        {
          name: "Gondwana",
          code: "GW",
        },
        {
          name: "Pannotia",
          code: "PN",
        },
      ]
    }}
  />);

  expect(content).toMatchSnapshot();
});

it("should render loading", () => {
  const content = shallow(<ContinentsList responseStatus="LOADING"/>);

  expect(content).toMatchSnapshot();
});

it("should render error", () => {
  const content = shallow(<ContinentsList responseStatus="ERROR"/>);

  expect(content).toMatchSnapshot();
});
