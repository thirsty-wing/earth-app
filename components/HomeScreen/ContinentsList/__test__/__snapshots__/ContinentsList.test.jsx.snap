// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error 1`] = `
<View
  style={
    {
      "flex": 1,
      "gap": 8,
    }
  }
>
  <Text>
    Something went wrong
  </Text>
</View>
`;

exports[`should render loading 1`] = `
<View
  style={
    {
      "flex": 1,
      "gap": 8,
    }
  }
>
  <ActivityIndicator
    size="large"
  />
</View>
`;

exports[`should render with continents 1`] = `
<View
  style={
    {
      "flex": 1,
      "gap": 8,
    }
  }
>
  <Memo(Pressable)
    onPress={[Function]}
    style={
      {
        "backgroundColor": "#115",
        "borderRadius": 16,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      Pangea
    </Text>
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      PG
    </Text>
  </Memo(Pressable)>
  <Memo(Pressable)
    onPress={[Function]}
    style={
      {
        "backgroundColor": "#115",
        "borderRadius": 16,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      Gondwana
    </Text>
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      GW
    </Text>
  </Memo(Pressable)>
  <Memo(Pressable)
    onPress={[Function]}
    style={
      {
        "backgroundColor": "#115",
        "borderRadius": 16,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      Pannotia
    </Text>
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      PN
    </Text>
  </Memo(Pressable)>
</View>
`;
